### PR TITLE
test(orchestration): ibc mocking utils

### LINKED ISSUE #9572: test support: ibc bridge mocks for @agoric/orchestration
## What is the Problem Being Solved?

In unit testing flows for `@agoric/orchestration`, the `EchoConectionKit` is used to mock network / ibc:

https://github.com/Agoric/agoric-sdk/blob/bc3bc6b6cf9f0a7017f5b17adf5e57c300f7e6fe/packages/orchestration/test/supports.ts#L107
https://github.com/Agoric/agoric-sdk/blob/ad29545174c43d481f1be7dbc6ce848c893a89e7/packages/orchestration/test/network-fakes.ts#L98-L108

This limits our ability to fully test contracts - we need to be able to simulate responses from a remote IBC Connection to see how a contract handles them.

In boot tests, we mocked some `Bridge.DIBC` events using responses for sim chains:
https://github.com/Agoric/agoric-sdk/blob/42b2e7ed09b2022a59bfddef7fda572e00fca5f5/packages/boot/tools/supports.ts#L358-L406

## Description of the Design

Provide a similar entry point for IBC mocks in @agoric/orchestration unit testing.

### PR DESCRIPTION
**Pull request by 0xpatrickdev**
closes: #9572 

## Description

Provides testing utilities for simulating IBC bridge events and packets in `@agoric/orchestration` .

### CODE REVIEW

File: packages/orchestration/test/network-fakes.test.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+test('network fakes - port allocator', async t => {
+  const { portAllocator } = t.context;
+  const customPort = await E(portAllocator).allocateCustomIBCPort('test-port');
+  t.is(await E(customPort).getLocalAddress(), '/ibc-port/custom-test-port');
+});
+
+test('network fakes - ibc connection', async t => {
+  const { portAllocator } = t.context;
+
+  // allocate ICA controller port and connection to remote chain
```
Comment (0xpatrickdev) [inline packages/orchestration/test/network-fakes.test.ts]:
```suggestion
  // allocate ICA controller port and connect to remote chain
```

File: packages/orchestration/test/network-fakes.ts [hunk 2]
```diff
@@ -1,116 +1,274 @@
...
+    setMockAck: (mockAckMap: Record<string, string>) => void;
+    setAddressPrefix: (addressPrefix: string) => void;
+  }
+> => {
+  let bridgeHandler: any;
+  /**
+   * Intended to mock an individual account's sequence, but is global for all
+   * accounts.
+   * XXX teach this about IBCConnections and store sequence on a
+   * per-connection basis.
```
Comment (0xpatrickdev) [inline packages/orchestration/test/network-fakes.ts]:
```suggestion
   * per-channel basis.
```

File: packages/cosmic-proto/src/helpers.ts [hunk 3]
```diff
@@ -103,7 +103,7 @@ export const typeUrlToGrpcPath = (typeUrl: Any['typeUrl']) => {
 type RequestQueryOpts = Partial<Omit<RequestQuery, 'path' | 'data'>>;
 
 export const toRequestQueryJson = (
-  msg: QueryBalanceRequestProtoMsg,
+  msg: { value: Uint8Array; typeUrl: string },
```
Comment (turadg) [inline packages/cosmic-proto/src/helpers.ts]: The type described here is `Any` https://github.com/Agoric/agoric-sdk/blob/4c51569c9814fba63865c6de46e4d67497e4be0a/packages/cosmic-proto/src/codegen/google/protobuf/any.ts#L117-L120
```suggestion
  msg: Any,
```

Please maintain use of aliases as documentation of what these types are, that they aren't ad-hoc.

Though I'd encourage keeping `QueryBalanceRequestProtoMsg` since that's the only query we yet support. We can add a union for other messages. I think you only had to make this Any because `buildQueryPacketString` didn't constrain its param, and I think it would be better to instead use `QueryBalanceRequestProtoMsg` in both places as a form of documentation. It doesn't affect future compatibility because it doesn't affect runtime.

If as we add more request types the union of them becomes unwieldy we could define a message type using `RequestTypeUrl`

It you don't use `QueryBalanceRequestProtoMsg` remember to remove its import.

File: packages/orchestration/src/utils/packet.js [hunk 4]
```diff
@@ -108,7 +108,7 @@ export function parseQueryPacket(response) {
   return harden(
     responses.map(resp => ({
       ...resp,
-      height: String(resp.index),
+      height: String(resp.height),
```
Comment (turadg) [inline packages/orchestration/src/utils/packet.js]: incidentally, I'm surprised there isn't a function already to do this encoding.

File: packages/orchestration/test/examples/stake-ica.contract.test.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+import path from 'path';
+import { makeNotifierFromSubscriber } from '@agoric/notifier';
+import type { Installation } from '@agoric/zoe/src/zoeService/utils.js';
+import {
+  QueryBalanceRequest,
+  QueryBalanceResponse,
+} from '@agoric/cosmic-proto/cosmos/bank/v1beta1/query.js';
+import { commonSetup } from '../supports.js';
+import { type StakeIcaTerms } from '../../src/examples/stakeIca.contract.js';
+import chainInfo from '../../src/fetched-chain-info.js';
```
Comment (turadg) [inline packages/orchestration/test/examples/stake-ica.contract.test.ts]: please name `fetchedChainInfo` to match its module and make more clear the source when reading within `getChainTerms`

File: packages/orchestration/test/examples/stake-ica.contract.test.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    await E(ibcBridge).setMockAck({
+      [buildQueryPacketString([
+        QueryBalanceRequest.toProtoMsg({
+          address: chainAddress.address,
+          denom: 'uosmo',
+        }),
+      ])]: buildQueryResponseString(QueryBalanceResponse, {
+        balance: { amount: '0', denom: 'uosmo' },
+      }),
+    });
```
Comment (turadg) [inline packages/orchestration/test/examples/stake-ica.contract.test.ts]: this is hard to parse visually. consider pulling out the key and maybe the value into named bindings

File: packages/orchestration/test/ibc-mocks.test.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    buildMsgResponseString(MsgDelegateResponse, {}),
+    'eyJyZXN1bHQiOiJFaTBLS3k5amIzTnRiM011YzNSaGEybHVaeTUyTVdKbGRHRXhMazF6WjBSbGJHVm5ZWFJsVW1WemNHOXVjMlU9In0=',
+  );
+});
+
+test('ibc-mocks - buildMsgErrorString matches observed values in e2e testing', t => {
+  t.is(
+    buildMsgErrorString(),
+    'eyJlcnJvciI6IkFCQ0kgY29kZTogNTogZXJyb3IgaGFuZGxpbmcgcGFja2V0OiBzZWUgZXZlbnRzIGZvciBkZXRhaWxzIn0=',
+    'ABCI code: 5: error handling packet',
```
Comment (turadg) [inline packages/orchestration/test/ibc-mocks.test.ts:29]: what's this test message meant to convey?

File: packages/orchestration/test/network-fakes.ts [hunk 8]
```diff
@@ -1,116 +1,274 @@
...
+
+  acknowledgementPacket: (
+    obj: IBCMethod<'sendPacket'>,
+    opts: { sequence: number; acknowledgement: string },
+  ): IBCEvent<'acknowledgementPacket'> => {
+    const { sequence, acknowledgement } = opts;
+    return {
+      acknowledgement,
+      blockHeight: 289,
+      blockTime: 1712180320,
```
Comment (turadg) [inline packages/orchestration/test/network-fakes.ts:103]: could these be random or incrementing? I don't see any testing of these particular values

File: packages/orchestration/test/service.test.ts [hunk 9]
```diff
@@ -44,20 +52,18 @@ test('makeICQConnection returns an ICQConnection', async t => {
...
-      ]),
+  const [result] = await E(icqConnection).query([
+    toRequestQueryJson(
+      QueryBalanceRequest.toProtoMsg({
+        address: 'cosmos1test',
+        denom: 'uatom',
+      }),
     ),
-    { message: /\\"data\\":\\"(.*)\\"memo\\":\\"\\"/ },
-    'TODO do not use echo connection',
```
Comment (turadg) [inline packages/orchestration/test/service.test.ts:59]: 👏

File: packages/vats/src/types.d.ts [hunk 10]
```diff
@@ -94,8 +94,7 @@ export type BridgeHandler = {
   fromBridge: (obj: any) => Promise<unknown>;
 };
 
-/** An object which handles messages for a specific bridge */
-export type ScopedBridgeManager<BridgeId extends BridgeIdValue> = Guarded<{
+export type ScopedBridgeMangerMethods<BridgeId extends BridgeIdValue> = {
```
Comment (turadg) [inline packages/vats/src/types.d.ts]: s/Manger/Manager

**Review by turadg — COMMENTED** [review]
This will greatly help our package testing coverage. I wonder about how we can DRY with RunUtils tests à la https://github.com/Agoric/agoric-sdk/issues/8963. I also wonder about how to improve or at least maintain fidelity of these mocks.

File: packages/orchestration/test/ibc-mocks.test.ts (hunk 7 above)
Comment (0xpatrickdev) [reply packages/orchestration/test/ibc-mocks.test.ts:29]: On our current version of `ibc-go`, it seems the error reported is module error code _without the corresponding codespace_. Codespace is how we'd determine if error code 5 means **sdk/5** _insufficient funds_, or or **bank/5** _send transactions are disabled_.
It seems like we may not be able to make much sense of these until we are on `ibc-go@^8.3.x`, but need to consider an interim solution. Here is a ticket to capture the issue: https://github.com/Agoric/agoric-sdk/issues/9629

File: packages/orchestration/test/examples/stake-ica.contract.test.ts (hunk 6 above)
Comment (0xpatrickdev) [reply packages/orchestration/test/examples/stake-ica.contract.test.ts]: I've made an edit to enhance readability, but don't expect this to be the final abstraction. We might limit verbosity with something like:
```js
mocks.delegate((reqPartial, respPartial, [error])  => [ReqString, RespString])
```
that can be massaged into a record.

We also might want to do do these instantiations in `test.before()`, so the actual test code isn't cluttered.

File: packages/cosmic-proto/src/helpers.ts (hunk 3 above)
Comment (0xpatrickdev) [reply packages/cosmic-proto/src/helpers.ts]: The `Any` suggestion is great, thank you. I intended to leverage existing types as much as possible but missed this one. I'm less convinced we need to include `QueryBalanceRequestProtoMsg`, but I've included it in the union anyway.

In somewhat related code I chose to type out `Encoder`. If there's something we can instead import from telescope, that'd be great. We needed something that accepted generic params so we can supply any `*Msg` and get type inference for the second argument (`Partial<T>`) of `buildMsgResponseString` and `buildQueryResponseString`.

File: packages/orchestration/test/examples/stake-ica.contract.test.ts (hunk 5 above)
Comment (0xpatrickdev) [reply packages/orchestration/test/examples/stake-ica.contract.test.ts]: Rename done and included `fetchChainInfo` as an optional second argument typed as `Record<string, CosmosChainInfo>` - `ChainInfo` was giving a hard time and this is a cosmos-specific contract

File: packages/orchestration/test/network-fakes.ts (hunk 8 above)
Comment (0xpatrickdev) [reply packages/orchestration/test/network-fakes.ts:103]: They could be, but yes, we are not testing anything against `blockTime` or `blockHeight` since application code does not currently rely on them.

These values seems like they could be useful for testing packet timeouts, but I suspect we can cover that path by using `buildMsgErrorString` and implicitly throwing a timeout error through the testing utils.

For testing something like whether `IBCMsgTransferOptions` are honored/correctly specified for `.transfer()`, e2e tests with sim chains seem like a better place for this.

File: packages/orchestration/src/utils/packet.js (hunk 4 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/utils/packet.js]: Good callout - see [1962218](https://github.com/Agoric/agoric-sdk/pull/9628/commits/1962218d7536915cbe10ef29e2efb306acc3426d)

**Comment by 0xpatrickdev** [comment]
> I wonder about how we can DRY with RunUtils tests à la https://github.com/Agoric/agoric-sdk/issues/8963.

Agree. Something at least directionally helpful would be to have `ibc-mocks` live in one place (`@agoric/orchestration`, i think) and have `package/boot` depend on it:
https://github.com/Agoric/agoric-sdk/blob/26c8b3daaf4f38ddcf0e0bb25945a1af89302504/packages/boot/tools/ibc/mocks.js

This was something considered but I recalled a suggestion to the effect of ~"every package should be responsible for it's own testing setup". Mocks / fixtures seems like they could be an exception to this, though.

LMK if you'd like to see that in this PR.


> I also wonder about how to improve or at least maintain fidelity of these mocks.

Agree, interested to hear more thoughts on this.

File: packages/orchestration/src/utils/packet.js [hunk 11]
```diff
@@ -105,14 +107,6 @@ export function parseQueryPacket(response) {
...
-  return harden(
-    responses.map(resp => ({
-      ...resp,
-      height: String(resp.height),
-      index: String(resp.index),
-      key: encodeBase64(resp.key),
-      value: encodeBase64(resp.value),
-    })),
-  );
+  return harden(responses.map(ResponseQuery.toJSON));
```
Comment (turadg) [inline packages/orchestration/src/utils/packet.js:110]: nice

File: packages/orchestration/test/examples/stake-ica.contract.test.ts [hunk 12]
```diff
@@ -104,16 +111,19 @@ test('makeAccount, getAddress, getBalances, getBalance', async t => {
     t.regex(chainAddress.address, /osmo1/);
     t.like(chainAddress, { chainId: 'osmosis-1' });
 
-    await E(ibcBridge).setMockAck({
-      [buildQueryPacketString([
+    const buildMocks = () => {
```
Comment (turadg) [inline packages/orchestration/test/examples/stake-ica.contract.test.ts:114]: good thinking to encapsulate the temporary values. another option is an IIFE

**Review by turadg — APPROVED** [review]
